cluster/health-monitor.sh

```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o nounset
+set -o pipefail
+
+# CRICTL is the path of crictl
+CRICTL=${CRICTL:-"crictl"}
+# COMMAND_TIMEOUT is the timeout for the health check command.
+COMMAND_TIMEOUT=${COMMAND_TIMEOUT:-60}
+# CHECK_PERIOD is the health check period.
+CHECK_PERIOD=${CHECK_PERIOD:-10}
+# SLEEP_SECONDS is the time to sleep after killing cri-containerd
+# and containerd.
+SLEEP_SECONDS=${SLEEP_SECONDS:-120}
+
+while true; do
+  # Use crictl sandboxes because it requires both containerd and
+  # cri-containerd to be working.
+  if ! timeout ${COMMAND_TIMEOUT} ${CRICTL} sandboxes > /dev/null; then
```

Comment:
prefer using `version` for health monitor over list `sandboxes`.. could be 100pods.  Version should do the same function but with less memory usage.

Reply:
That said this is a fairly expensive loop, every ten seconds creates a new instance of crictl, does all the init work of connecting to cri-containerd, grabs a list of all the sandboxes... go to sleep 10sec.

Let's open an issue with CRI for a health check api.. Off the top of my head here's how I would do it:  I'd add a ping / pong service api to the CRI image and runtime services.  I'd add a health check command to crictl that would "ping" the named CRI service every n-seconds, and return with a timeout error when / if the pong is not received before the time out. I'd rather see the pings coming from kubelet than a crictl loop, but I can see that kubelet would not necessarily know how to init cri-containerd/containerd. That's another thing I'd like to fix :-) IMO there should be some way for kubelet to manage the CRI runtime engines.

Reply:
>prefer using version for health monitor over list sandboxes.. could be 100pods. Version should do the same function but with less memory usage.

`version` can not seem to prove whether `containerd` is working normally or not. I think it better to add a health check api for CRI. :)

Reply:
Yeah not sure if the list of pods is proof either :-) A health check api would leave it up to the daemon to pong or otherwise notify via some eventing system when it believes it's important functions are still working which could be based on internal state flags (time stamp values) or whatever health system the daemon wants to use.

Reply:
@mikebrow We do have a `Status` and a `Version` function in CRI. However, it's not sufficient to check whether the cri-containerd/containerd is healthy.

For docker we usually encounter the issue that `docker version` still works, but `docker ps` stucks forever. That's why we are using `docker ps` in the current health monitor. https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/health-monitor.sh#L29

Kubelet does list every second, and we do have benchmark for this behavior, the overhead is acceptable.

Reply:
Filed an issue https://github.com/kubernetes-incubator/cri-containerd/issues/491 for this.